pyrsistent/_pclass.py

```diff
@@ -223,7 +220,29 @@ def __getitem__(self, item):
         return self._pclass_evolver_data[item]
 
     def set(self, key, value):
-        if self._pclass_evolver_data.get(key, _MISSING_VALUE) is not value:
+        update = False
+
+        current_value = self._pclass_evolver_data.get(key, _MISSING_VALUE)
+        try:
+            if hash(current_value) == hash(value) and current_value == value:
```

Comment:
I might be missing something, but it seems like it's probably unnecessary to compare the hashes of the old and new values.
Also I'd prefer if the try...except was narrower.

I came up with this implementation, which still passes the tests.

```
    def set(self, key, value):
        current_value = self._pclass_evolver_data.get(key, _MISSING_VALUE)

        if current_value is value:
            # Return early if the new value *is* the old value.
            return self
        else:
            update = True

        try:
            hash(value)
        except TypeError:
            pass
        else:
            if current_value == value:
                # If the value is hashable then it is immutable. If the value
                # is the same for an immutable object, then we do not need to
                # update.
                update = False

        if update:
            self._pclass_evolver_data[key] = value
            self._pclass_evolver_data_is_dirty = True

        return self

```